### PR TITLE
topology: pre-processor: Add support for object arrays

### PR DESCRIPTION
**Pull request by ranj063**
Introduce a new keyword, "CombineArrays" to instantiate multiple nodes
of the type specified. For example:

CombineArrays.Object.Base.input_audio_format [
	{
		in_rate [
			8000
			16000
			48000
		]
		in_bit_depth [
			32
		]
		in_valid_bit_depth [
			24
			32
		]
	}
]

This would be expanded into 6 objects with the rate, bit_depth and
valid_bit_depth combinations of the arrays of values above.

Object.Base.input_audio_format [
	{
		in_rate 8000
		in_bit_depth 32
		in_valid_bit_depth 32
	}
	{
		in_rate 16000
		in_bit_depth 32
		in_valid_bit_depth 32
	}
	{
		in_rate 48000
		in_bit_depth 32
		in_valid_bit_depth 32
	}
	{
		in_rate 8000
		in_bit_depth 32
		in_valid_bit_depth 24
	}
	{
		in_rate 16000
		in_bit_depth 32
		in_valid_bit_depth 24
	}
	{
		in_rate 48000
		in_bit_depth 32
		in_valid_bit_depth 24
	}
]

The CombineArrays definition is an array so that multiple combinations can
be specified in order to deal with only valid combinations. For example,
16-bit bit_depth is only valid with 16-bit valid_bit_depth. So if we
also want to add those combinations with the multiple rates, the
definition would look like:

CombineArrays.Object.Base.input_audio_format [
        {
                in_rate [
                        8000
                        16000
                        48000
                ]
                in_bit_depth [
                        32
                ]
                in_valid_bit_depth [
                        24
                        32
                ]
        }
        {
                in_rate [
                        8000
                        16000
                        48000
                ]
                in_bit_depth [
                        16
                ]
                in_valid_bit_depth [
                        16
                ]
        }
]

### CODE REVIEW

**Review by singalsu — APPROVED** [review]
Thanks, this is a good improvement for editing ease, saves time and avoids mistakes with very long format .conf blurb. I just ran an aplay & arecord test with topology that was generated with this patch and it worked correctly.

File: topology/pre-processor.c [hunk 1]
```diff
@@ -637,6 +637,332 @@ static int pre_process_includes_all(struct tplg_pre_processor *tplg_pp, snd_conf
...
+
+static int pre_process_object_array(struct tplg_pre_processor *tplg_pp, snd_config_t *top)
+{
+	snd_config_iterator_t i, next;
+	snd_config_t *objects;
+	const char *top_id;
+	int object_count;
+	int ret;
+
+	ret = snd_config_search(top, "Objects", &objects);
```
[the remark this answers is not in the document]
Comment (ranj063) [reply topology/pre-processor.c]: Thanks @singalsu . I've updated with the new keyword "ObjectArray" now

**Comment by ranj063** [comment]
@perexg any thoughts about this feature and it's implementation?

File: topology/pre-processor.c [hunk 2]
```diff
@@ -689,6 +1012,13 @@ int pre_process(struct tplg_pre_processor *tplg_pp, char *config, size_t config_
 		fprintf(stderr, "Failed to process conditional includes in input config\n");
 		goto err;
 	}
+
+	/* expand object arrays */
```
Comment (kv2019i) [inline topology/pre-processor.c:1016]: @ranj063  The name "ObjectArray" is quite misleading as this is not an array of objects, but rather a combinatorial initializer, something like Python itertools.combinations https://docs.python.org/3/library/itertools.html#itertools.combinations

Maybe ObjectArrayCombinations ?

Comment (perexg) [reply topology/pre-processor.c:1016]: Or `ObjectCombineArrays` . I agree that the identifier is misleading.

**Comment by perexg** [comment]
I am basically ok with the change, but it seems like a generic config tree change. Why to restrict this only to the Object subtree? I would create `CombineArrays` block as:

```
CombineArrays.Object.Base.input_audio_format  [
   ...
]
```

**Comment by ranj063** [comment]
> I am basically ok with the change, but it seems like a generic config tree change. Why to restrict this only to the Object subtree? I would create `CombineArrays` block as:
> 
> ```
> CombineArrays.Object.Base.input_audio_format  [
>    ...
> ]
> ```

@perexg sorry for the delay. I have updated the PR to use "CombineArrays" as the new generic keyword which expands the array into multiple conf nodes with the specified id.

**Pull request closed without merging**